types/msgservice/msg_service.go

```diff
@@ -21,7 +23,7 @@ func RegisterMsgServiceDesc(registry codectypes.InterfaceRegistry, sd *grpc.Serv
 		// NOTE: This is how we pull the concrete request type for each handler for registering in the InterfaceRegistry.
 		// This approach is maybe a bit hacky, but less hacky than reflecting on the handler object itself.
 		// We use a no-op interceptor to avoid actually calling into the handler itself.
-		_, _ = methodHandler(nil, context.Background(), func(i interface{}) error {
+		r, _ := methodHandler(nil, context.Background(), func(i interface{}) error {
```

Comment:
`r` is unfortunately nil here.

Any ideas how I can get the response type?
- do we actually need to run the handler (hope not)?
- or reflection?

Reply:
> 

this should work:
```go
package msgservice

import (
	"bytes"
	"compress/gzip"
	sdk "github.com/cosmos/cosmos-sdk/types"
	"github.com/gogo/protobuf/proto"
	proto2 "google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/reflect/protodesc"
	"google.golang.org/protobuf/reflect/protoreflect"
	"google.golang.org/protobuf/types/descriptorpb"
	"io/ioutil"
	"log"
	"reflect"

	"google.golang.org/grpc"

	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
)

// RegisterMsgServiceDesc registers all type_urls from Msg services described
// in `sd` into the registry.
func RegisterMsgServiceDesc(registry codectypes.InterfaceRegistry, sd *grpc.ServiceDesc) {
	fdBytesZipped := unzip(proto.FileDescriptor(sd.Metadata.(string)))
	if fdBytesZipped == nil {
		panic("oh no")
	}

	fdRaw := &descriptorpb.FileDescriptorProto{}
	err := proto2.Unmarshal(fdBytesZipped, fdRaw)
	if err != nil {
		panic(err)
	}
	fd, err := protodesc.FileOptions{
		AllowUnresolvable: true,
	}.New(fdRaw, nil)
	if err != nil {
		panic(err)
	}
	prefSd := fd.Services().ByName(protoreflect.FullName(sd.ServiceName).Name())
	for i := 0; i < prefSd.Methods().Len(); i++ {
		md := prefSd.Methods().Get(i)
		requestDesc := md.Input()
		responseDesc := md.Output()

		reqTyp := proto.MessageType(string(requestDesc.FullName()))
		respTyp := proto.MessageType(string(responseDesc.FullName()))
		// register here sdk.MSg and sdk.MsgResp
		log.Printf("%T", reflect.New(reqTyp).Elem().Interface().(sdk.Msg))
		log.Printf("%T", reflect.New(respTyp).Elem().Interface())
	}

}

func unzip(b []byte) []byte {
	if b == nil {
		return nil
	}
	r, err := gzip.NewReader(bytes.NewReader(b))
	if err != nil {
		panic(err)
	}

	unzipped, err := ioutil.ReadAll(r)
	if err != nil {
		panic(err)
	}

	return unzipped
}
```